tests/test_frame.py

```diff
@@ -1980,6 +1980,46 @@ def test_groupby_result() -> None:
         pass
 
 
+def test_groupby_result_for_scalar_indeces() -> None:
+    # GH 674
+    df = pd.DataFrame({"date": pd.date_range("2020-01-01", "2020-12-31"), "days": 1})
+    period_index = pd.PeriodIndex(df.date, freq="M")
+    iterator = df.groupby(period_index).__iter__()
+    assert_type(iterator, Iterator[Tuple[pd.Period, pd.DataFrame]])
+    index, value = next(iterator)
+    assert_type((index, value), Tuple[pd.Period, pd.DataFrame])
+
+    check(assert_type(index, pd.Period), pd.Period)
+    check(assert_type(value, pd.DataFrame), pd.DataFrame)
+
+    dt_index = pd.DatetimeIndex(df.date)
+    iterator2 = df.groupby(dt_index).__iter__()
+    assert_type(iterator2, Iterator[Tuple[Scalar, pd.DataFrame]])
+    index2, value2 = next(iterator2)
+    assert_type((index2, value2), Tuple[Scalar, pd.DataFrame])
+
+    check(assert_type(index2, Scalar), pd.Timestamp)
+    check(assert_type(value2, pd.DataFrame), pd.DataFrame)
```

Comment:
When I look at this sequence, since you know the `Scalar` will be a `Timestamp`, might as well create an overload for that argument and have `__iter__()` then return `tuple[pd.Timestamp, pd.DataFrame]`

This is also another reason to make it `Generic`

Reply:
If we're going to add a bunch of overloads anyways, shall I also make ones for the rest of the index types that are neither `Index` nor `MultipleIndex`?

Making `GroupByDataFrame` generic we could also be a little bit more careful about where we put the `Index` overload. Since we don't really know the `dtype` of the index we should probably just return `GroupByDataFrame[Any]`.

- `CategoricalIndex` is in the same boat as `Index`, since we don't really know the type of the categories, so that could be part of the same overload.
- `MultiIndex` in the overload that originally had `Index` should be fine, since the group identifier will be a tuple.
- `IntervalIndex` would need its own overload with `pd.Interval`, it's currently also not part of `Scalar` just like `pd.Period`.

Reply:
I did end up going with exactly that

Reply:
I like the path that you took with this